specs/_features/eip7732/beacon-chain.md

```diff
@@ -0,0 +1,716 @@
+    if voted_present != payload_was_present:
+        # Unset the flags in case they were set by an equivocating ptc attestation
+        proposer_penalty_numerator = 0
+        for index in indexed_payload_attestation.attesting_indices:
+            for flag_index, weight in enumerate(PARTICIPATION_FLAG_WEIGHTS):
+                if has_flag(epoch_participation[index], flag_index):
+                    epoch_participation[index] = remove_flag(epoch_participation[index], flag_index)
+                    proposer_penalty_numerator += get_base_reward(state, index) * weight
+        # Penalize the proposer
+        proposer_penalty = Gwei(2 * proposer_penalty_numerator // proposer_reward_denominator)
+        decrease_balance(state, proposer_index, proposer_penalty)
+        return
+
+    # Reward the proposer and set all the participation flags in case of correct attestations
+    proposer_reward_numerator = 0
+    for index in indexed_payload_attestation.attesting_indices:
+        for flag_index, weight in enumerate(PARTICIPATION_FLAG_WEIGHTS):
+            if not has_flag(epoch_participation[index], flag_index):
+                epoch_participation[index] = add_flag(epoch_participation[index], flag_index)
+                proposer_reward_numerator += get_base_reward(state, index) * weight
+
+    # Reward proposer
+    proposer_reward = Gwei(proposer_reward_numerator // proposer_reward_denominator)
+    increase_balance(state, proposer_index, proposer_reward)
+```
+
+#### Modified `process_execution_payload`
+
+##### New `verify_execution_payload_envelope_signature`
+
+```python
+def verify_execution_payload_envelope_signature(
+        state: BeaconState, signed_envelope: SignedExecutionPayloadEnvelope) -> bool:
+    builder = state.validators[signed_envelope.message.builder_index]
+    signing_root = compute_signing_root(signed_envelope.message, get_domain(state, DOMAIN_BEACON_BUILDER))
+    return bls.Verify(builder.pubkey, signing_root, signed_envelope.signature)
+```
+
+*Note*: `process_execution_payload` is now an independent check in state transition. It is called when importing a signed execution payload proposed by the builder of the current slot.
+
+```python
+def process_execution_payload(state: BeaconState, 
+                              signed_envelope: SignedExecutionPayloadEnvelope, 
+                              execution_engine: ExecutionEngine, verify: bool = True) -> None:
+    # Verify signature
+    if verify: 
+        assert verify_execution_payload_envelope_signature(state, signed_envelope)
+    envelope = signed_envelope.message
+    payload = envelope.payload
+    # Cache latest block header state root
+    previous_state_root = hash_tree_root(state)
+    if state.latest_block_header.state_root == Root():
+        state.latest_block_header.state_root = previous_state_root
+    
+    # Verify consistency with the beacon block
+    assert envelope.beacon_block_root == hash_tree_root(state.latest_block_header)
```

Comment:
This introduces an extra round trip because it means that `SignedExecutionPayloadEnvelope` can only be published after `SignedBeaconBlock` is available. For a pure consistency check, the `block_hash` check may possibly be sufficient.

Reply:
An envelope may also refer to same block hash with different block roots

Reply:
Why is that a problem? If multiple `BeaconBlock` refer to the same `ExecutionPayload`, why do we have to propagate the `ExecutionPayload` multiple times?

Note it's only possible with equivocation, because the `ExecutionPayload` hashes over the `timestamp` (so is based on `slot`)